src/network/netdev/batadv.c

```diff
@@ -0,0 +1,205 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <inttypes.h>
+
+#include "netdev/batadv.h"
+#include "network-internal.h"
+#include "networkd-manager.h"
+#include "netlink-util.h"
+#include <linux/genetlink.h>
+
+#include "fileio.h"
+#include "parse-util.h"
+#include "stdio-util.h"
+#include "string-util.h"
+
+static void batadv_init(NetDev *n) {
+        _unused_ Batadv *b;
+
+        b = BATADV(n);
+
+        /* Set defaults */
+        b->aggregation=true;
+        b->bridge_loop_avoidance=true;
+        b->distributed_arp_table=true;
+        b->fragmentation=true;
+        b->hop_penalty=true;
+}
+
+int config_parse_batadv_gateway_mode(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        Batadv *b = userdata;
+        int mode;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        mode = BATADV_GW_MODE_OFF;
+
+        if (streq(rvalue, "client"))
+                mode = BATADV_GW_MODE_CLIENT;
+
+        else if (streq(rvalue, "server"))
+                mode = BATADV_GW_MODE_SERVER;
+
+        else {
+                log_syntax(unit, LOG_WARNING, filename, line, 0,
+                          "Batadv gateway mode '%s' unknown. Ignoring.", rvalue);
+                return 0;
+        }
+
+        b->gateway_mode = mode;
+
+        return 0;
+}
+
+int config_parse_badadv_bandwidth (
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        uint64_t k;
+        uint64_t bandwidth;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        Batadv *b = userdata;
+
+        if (isempty(rvalue)) {
+                b->gateway_bandwidth_up = 0;
+
+                return 0;
+        }
+
+        r = parse_size(rvalue, 1000, &k);
+        if (r < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Failed to parse '%s=', ignoring assignment: %s",
+                           lvalue, rvalue);
+                return 0;
+        }
+
+        bandwidth = k/1000/100;
```

Comment:
should this really be rounded down? i.e. not rounded up? (DIV_ROUND_UP if so). 

I really don't know, i am just wondering

Reply:
hmm, if user specifies UINT64_MAX or so as size value this will resolve to 18446744073709551615 which is still substantially above UINT32_MAX, but the bandwitdh fields are actually 32bit. This should be caught and an error be generated.